docs/proposals/device-crd.md

```diff
@@ -0,0 +1,457 @@
+          type: string
+        value: OFF
+      expected:
+        metadata:
+          timestamp: '1550049403598'
+          type: string
+        value: OFF
+```
+
+### CRD API Group and Version Details
+The tables below summarize the group , kind and API version details for the CRDs
+
+* DeviceModel
+
+| Field                 | Description             |
+|-----------------------|-------------------------|
+|Group                  | devices.kubeedge.io     |
+|APIVersion             | v1alpha1                |
+|Kind                   | DeviceModel             |
+
+* DeviceInstance
+
+| Field                 | Description             |
+|-----------------------|-------------------------|
+|Group                  | devices.kubeedge.io     |
+|APIVersion             | v1alpha1                |
+|Kind                   | Device                  |
+
+### Device Model and Instance Types Definition
+
+```go
+// DeviceModelSpec defines the desired state of the device model
+type DeviceModelSpec struct {
+  // list of device properties
+	DeviceProperties []DeviceProperty              `json:"properties,omitempty"`
+  // list of device property visitors
+  DevicePropertyVisitors []DevicePropertyVisitor `json:"propertyVisitors,omitempty"`
+}
+
+// Device property
+type DeviceProperty struct {
+  // name of the device property
+	Name string                `json:"name,omitempty"`
+  // description of the device property
+	Description string         `json:"description,omitempty"`
+  // additional metadata about the device property like datatype, range, min, max etc.
+	Metadata map[string]string `json:"metadata,omitempty"`
+}
+
+// Device property visitor is the access mechanism to read/write a property
+type DevicePropertyVisitor struct {
+  // the protocol name used to access the property
+	ProtocolName string            `json:"protocolName,omitempty"`
+  // a list of property visitors describing how to access each property.
+	Visitors     []PropertyVisitor `json:"visitors,omitempty"`
+}
+
+// Device property visitor
+type PropertyVisitor struct {
+	// device property name
+  PropertyName        string            `json:"propertyName,omitempty"`
+  // device property visitor description
+  Description string                    `json:"description,omitempty"`
+  // additional metadata about the how to access the device property
+  Metadata    map[string]string         `json:"metadata,omitempty"`
+}
+
+// DeviceSpec defines the desired state of Device
+type DeviceSpec struct {
+  // device model used to create the device
+	DeviceModel *core.ObjectReference `json:"deviceModel,omitempty"`
+  // protocol configuration to talk to the device
+	Protocol    Protocol              `json:"protocol,omitempty"`
+  // node to which the device belongs
+	NodeName    string                `json:"nodeName,omitempty"`
+}
+
+// Protocol configuration
+type Protocol struct {
+  // protocol name e.g. ble , modbus, opcua etc.
+  Name string              `json:"name,omitempty"`
```

Comment:
Probably make protocol name to pre-defined string, or use similar design with kubernetes pod.spec.volumes to make easier enabling validation for protocol config

Reply:
Like what Kevin suggested. It is like a predefined enum.

Reply:
I have introduced constants for the protocol types. Thanks.